FastAPI routes for tool data tables

davelopez opened the pull request:
Ref #10889

## Summary
- Some additional functional tests to ensure a 404 error is returned as expected
- A test for the `reload` endpoint
- Created a [`ToolDataManager`](https://github.com/galaxyproject/galaxy/commit/372f6d6ce4dd3991a8ff736da25da71109216852#diff-37c16e129a9fcbfa5affe95b516f597fb0981ba345e5344d07dbd329a2bcaee7R10) to share as much logic as possible between the legacy and FastAPI controllers
- Added `aiofiles` as [dependency](https://github.com/galaxyproject/galaxy/commit/cdd9c7b6a777c440e01d236928a9343f2859f33a) to deal with `FileResponse` when downloading files using FastAPI

## Some additional comments
- Again, I am not 100% happy with the naming of classes, models, etc. that I have used. Please make suggestions if you feel the same :)
- The current `reload` action is modelled with a `GET`. Should it be a `POST`/`PUT` returning 200 instead of a `GET` operation since it may alter the data table? I am trying to figure out what is best practice, but changing it will break the legacy compatibility so probably not worth it.
- The pydantic `BaseModel` classes already have a `fields` property, so, an alias must be used to avoid conflicts with the `fields` attribute of [`ToolDataDetails`](https://github.com/galaxyproject/galaxy/compare/dev...davelopez:fastapi-migration-tool_data?expand=1#diff-01ebf253d063e9aaae129695c4ca04a8845deb143f290d78923a164f96190d1aR47) and [`ToolDataField`](https://github.com/galaxyproject/galaxy/compare/dev...davelopez:fastapi-migration-tool_data?expand=1#diff-01ebf253d063e9aaae129695c4ca04a8845deb143f290d78923a164f96190d1aR69) models. I have added some comments in the legacy endpoints that need to serialize the response explicitly using an alias with [`.dict(by_alias=True)`](https://github.com/galaxyproject/galaxy/compare/dev...davelopez:fastapi-migration-tool_data?expand=1#diff-db52a8a55c79b7d74d6dd77a5e6ddc3dbc5e51049067cf4fa247874c29e891b5R163)

davelopez committed:
Create ToolDataManager and initial schema models

davelopez committed:
Add index endpoint reusing ToolDataManager

davelopez committed:
Add show endpoint

davelopez committed:
Use InferringRouter in tool_data router

And remove all response_model declarations

davelopez committed:
Fix show endpoint using alias

The ToolDataDetails model can not have an attribute called 'fields' since it shadows the one on the pydantic BaseModel. Using an alias for the field solves the problem.

davelopez committed:
Add more tests

Additional tests to check 404 errors and reload endpoint

davelopez committed:
Refactor move show_field and reload to manager

davelopez committed:
Add show_field and reload enpoints

davelopez committed:
Add get_field_file_path method in manager

This will be shared by the new and the legacy API method

davelopez committed:
Add 404 test for download file

davelopez committed:
Migrate download_field_file endpoint to FastAPI

davelopez committed:
Add aiofiles dependency

mvdbeek commented on lib/galaxy/webapps/galaxy/api/tool_data.py:
```diff
@@ -1,21 +1,143 @@
+    ):
+        """Download a file associated with the data table field."""
+        path = self.tool_data_manager.get_field_file_path(table_name, field_name, file_name)
+        return FileResponse(str(path))
```
I think we'll want to do what we do in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/web/framework/base.py#L484 somehow.
Maybe doing that as a middleware would be the best approach ? Or maybe as a dependency ?

davelopez replied:
@mvdbeek 
I have been playing around with something like:
````python
def add_send_file_middleware(app: FastAPI, gx_config=Depends(?)) -> None:
    @app.middleware("http")
    async def add_send_file_header(request: Request, call_next) -> Response:
        response = await call_next(request)
        if isinstance(response, StreamingResponse):
            if gx_config.nginx_x_accel_redirect_base:
                response.headers['X-Accel-Redirect'] = path_to_file
            elif gx_config.apache_xsendfile:
                response.headers['X-Sendfile'] = path_to_file
        return response
````
But I had some troubles figuring out how to get the config properly. Maybe when we merge #11180 we can look into it again? I will put it in my TODO list, promise :)

jmchilton replied:
The test data for these tools is relatively small so it is probably less important this works with this controller, but when we get to the datasets controller and such we will definitely need that working I think.

jmchilton commented on lib/galaxy/webapps/galaxy/api/tool_data.py:
```diff
@@ -24,12 +146,25 @@ def index(self, trans, **kwds):
+
+        :type   id:     str
+        :param  id:     the id of the data table
+        """
```
I wouldn't worry about adding docs to API endpoints we're immediately deprecating in the same PR 😁.

davelopez replied:
You are right 😅

jmchilton commented on lib/galaxy/managers/tool_data.py:
```diff
@@ -0,0 +1,81 @@
+from typing import Dict
+
+from galaxy import exceptions
+from galaxy.app import UniverseApplication
```
Just merged https://github.com/galaxyproject/galaxy/pull/11142/files to provide a better way of doing this I think. Any chance you can rebase and import ``galaxy.app.structured_app.StructuredApp`` instead for stronger typing and better dependency management.

davelopez replied:
Sure! I will do it. Let's see if this time I don't destroy the whole galaxy with my rebase attempt 😆

davelopez replied:
Done!

jmchilton commented on lib/galaxy/managers/tool_data.py:
```diff
@@ -0,0 +1,81 @@
+        return self._app.tool_data_tables.data_tables
+
+    @property
+    def queue_worker(self) -> GalaxyQueueWorker:
```
If it isn't meant to be part of the public interface of the class I guess we should prefix the name with ``_``. I think this should probably be ``_queue_worker``.

jmchilton replied:
Actually if you just use ``self._app.queue_worker`` in the one place this is used you get all the typing benefit with the new StructureApp approach but you don't need to import ``GalaxyQueueWorker`` and make it explicit like this. I could go either way here though - this helper wouldn't be my preference but making this a little more explicit here might improve readability?

davelopez replied:
Absolutely!, I missed it, thanks for the catch :)

davelopez replied:
I like your suggestion of using `self._app.queue_worker` and remove the attribute and the import completely. It is used only once and it's quite readable and less code.

jmchilton commented on lib/galaxy/webapps/galaxy/api/tool_data.py:
```diff
@@ -1,21 +1,143 @@
+        summary="Reloads a tool data table",
+        response_description="A description of the reloaded data table and its content",
+        dependencies=[
+            AdminUserRequired
```
Ooooo, that is kind of nice!

davelopez committed:
Address some comments in the review

davelopez committed:
Import StructuredApp instead of UniverseApplication

jdavcs commented on lib/galaxy/webapps/galaxy/api/tool_data.py:
```diff
@@ -1,22 +1,144 @@
+
+router = APIRouter(tags=['tool data'])
+
+AdminUserRequired = Depends(get_admin_user)
```
This little change makes the code read so much more intuitive! `Depends=get_admin_user` - you assume you need the return value, then see it's in the route decorator - so you don't, and then finally see its purpose. `Depends=AdminUserRequired` simply doesn't raise any questions at all. Nice!

davelopez replied:
Glad you like it :)

davelopez committed:
Add tests for delete errors

davelopez committed:
Add delete method to manager

davelopez committed:
Add delete endpoint

davelopez committed:
Minor doc changes and cleanup

davelopez commented on lib/galaxy_test/api/test_tool_data.py:
```diff
@@ -44,6 +44,43 @@ def test_download_field_file(self):
+
     # Following test case rendered invalid by the fix in
     # https://github.com/galaxyproject/galaxy/commit/48f77dc742acf01ddbafafcc4634e69378f1f020#diff-bfb557a99c1f7d646d4968d8d680b885R154.
     # TODO: Restore the test case when test framework allows actions from
```
Any chance this can be tested now?
I have tried to uncomment the test and fix it a bit, but ended up with the error message: `Job 1 has no user and no session.` so I assume it is still not possible to run it.
Just to be sure I have tested the `DELETE` endpoint manually and it seems to work, but it would be great if we can test it properly :)

jmchilton replied:
I'm working on it, should be possible.

jmchilton replied:
There are two fixes in https://github.com/jmchilton/galaxy/pull/new/tool_data_fixes. Want to cherry-pick these into this branch?

davelopez replied:
Done! Thank you so much @jmchilton

davelopez replied:
I will try to create the integration tests for this :)

davelopez replied:
@jmchilton I moved the `test_delete_entry` to an integration test [here](https://github.com/galaxyproject/galaxy/pull/11167/commits/c11eca20d6ddcf6dc97b870280f0b167e1ec903a). Hope that is what you had in mind.

jmchilton replied:
Beautiful!

jmchilton committed:
Fix bug where HERE in tool data tables breaks down after reload.

jmchilton committed:
Allow execution of tool data deletion test case.

It ultimately needs to be an integration test since it modifies files under git control as is, but it can be hacked around for now to test this functionality.

GALAXY_TEST_INCLUDE_API_CONFIG_MOD_TESTS=1 pytest -s lib/galaxy_test/api/test_tool_data.py::AdminToolDataApiTestCase

davelopez committed:
Update test_delete_entry to reduce mutability of config files

Delete the last entry intead of the first one (the original). This should reduce the mutation probability of the server config files.

davelopez committed:
Move test_delete_entry to integration tests
